### PR TITLE
[framework] Adds sui::pay module for wallet functionality

### PR DESCRIPTION
**Pull request by damirka**
This PR suggests splitting functionality between the core `Coin` and a set of handy functions for Coin management. I hope that in future wallets will build their own wallet-modules that will suit their goals perfectly; this is a first step to show that this is a possibility.

Another reasons for this change:

1. Protects Coin module from further changes (and bloating) by moving coin management functionality to `sui::pay`
2. Solves the issue with `coin::split` not returning a value which is an important step towards `&mut Coin`

- naming is something I'm not sure about @sblackshear @tnowacki 

> Tests and the rest of implementation is left intact until we settle on whether this is something we want to have.

### CODE REVIEW

File: crates/sui-framework/sources/wallet.move [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        while (i < n - 1) {
+            vector::push_back(&mut vec, coin::split(self, split_amount, ctx));
+            i = i + 1;
+        };
+        vec
+    }
+
+    /// Split coin `self` into `n` coins with equal balances. If the balance is
+    /// not evenly divisible by `n`, the remainder is left in `self`.
+    public entry fun split_n<T>(self: &mut Coin<T>, n: u64, ctx: &mut TxContext) {
```
Comment (sblackshear) [inline crates/sui-framework/sources/wallet.move:54]: `split_n_to_vec` and `split_n` are different enough from `split` that I think we should use a different term. Perhaps we can "divide" instead of "split_n" (since it is dividing the coin into equal parts)?

**Review by sblackshear — COMMENTED** [review]
Suggested rule of thumb for when a function goes into `coin` vs `wallet`:
- If a function returns a `Coin` or `vector` of `Coin`'s, it belongs in `coin`.
- If it does a `transfer`, it belongs in `wallet`

This seems easy to apply consistently, explain in docs, and I think it will also help with naming--e.g., instead of being called `x_and_transfer`, the `coin` and `wallet function can both be called `x`.

**Comment by sblackshear** [comment]
On naming: I would suggest `payment` or `pay` rather than `wallet`. The reason:
- I think these API's will be used for modules that implement payment logic, not just by external wallets
- I suspect that we will end up including some "smart contract wallets" (e.g., multisig, social recovery) in the Sui Framework, and it would be nice to reserve the "wallet" name for those.

File: crates/sui-framework/sources/coin.move [hunk 2]
```diff
@@ -119,7 +112,7 @@ module sui::coin {
         balance::join(balance, into_balance(coin));
     }
 
-    // === Functionality for Coin<T> holders ===
+    // === Base Coin functionality ===
```
Comment (gegaowp) [inline crates/sui-framework/sources/coin.move:115]: a naive q, for public functions related to split and join, why not have one function that takes a vector of coins and transforms them into a new list of coins with specified amounts, like what we do with `pay` in native [Sui transaction](https://github.com/gegaowp/sui/blob/602c1b266a9eea04ab3b06d1a5d4569b5c4ed195/crates/sui-types/src/messages.rs#L124)? 
In that sense, other modules can always count on this one function to do all all split and merging. The idea is very similar @oxade 's transformer change? https://github.com/MystenLabs/sui/pull/4528/files
cc @sblackshear

Comment (sblackshear) [reply crates/sui-framework/sources/coin.move:115]: I think this would be a very useful function to add this this new `wallet` module.

I do think that's it's separately useful to have `split`/`join` for singleton coins, only because it's somewhat cumbersome to put coins into vectors + take them out (and you will eventually have to take them out in order to transfer or spend them).

Comment (oxade) [reply crates/sui-framework/sources/coin.move:115]: Yes we plan to add this function in future, which is why I paused transformer.
Adding this requires a bit more thought (especially if we want to reuse the Pay API logic by putting common code in Rust/natives).
I'd rather do that as a separate more targeted PR.

**Comment by damirka** [comment]
> Suggested rule of thumb for when a function goes into `coin` vs `wallet`:
> 
> * If a function returns a `Coin` or `vector` of `Coin`'s, it belongs in `coin`.
> * If it does a `transfer`, it belongs in `wallet`
> 
> This seems easy to apply consistently, explain in docs, and I think it will also help with naming--e.g., instead of being called `x_and_transfer`, the `coin` and `wallet function can both be called `x`.
@sblackshear 

First I tried this approach but there are few cases such as `coin::keep` which I would personally love to keep in Coin and not require additional import for it. As for returning a `Coin` I believe that function like `divide_n` or, say, we add "split into specified amounts" should be kept in the `pay` / `wallet` module rather than in `coin` as these are too specific or "fancy".

I understand that it's not a clear distinction (although I'd prefer there was some) and suggest to follow common sense above else.

File: crates/sui-framework/sources/pay.move [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        let len = vector::length(&split_amounts);
+        while (i < len) {
+            split(self, *vector::borrow(&split_amounts, i), ctx);
+            i = i + 1;
+        };
+    }
+
+    /// Split coin `self` into `n` coins with equal balances. If the balance is
+    /// not evenly divisible by `n`, the remainder is left in `self`. Return
+    /// newly created coins.
```
Comment (amnn) [inline crates/sui-framework/sources/pay.move:41]: If I understand the implementation of this function, it splits `self` into `n - 1` new coins with equal balance, and leaves the remainder in `self` (which could be equal to the other coins if its value was divisible by `n`, but it could also be more).

To me, the documentation currently suggests something different:  That it produces `n` new coins, and `self` gets `0` value if it all divides cleanly or the remainder i.e. strictly less than `coin::value(self) / n` otherwise.

File: crates/sui-framework/sources/pay.move [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        };
+        vector::destroy_empty(vec);
+    }
+
+    /// Join everything in `coins` with `self`
+    public entry fun join_vec<T>(self: &mut Coin<T>, coins: vector<Coin<T>>) {
+        let i = 0;
+        let len = vector::length(&coins);
+        while (i < len) {
+            let coin = vector::remove(&mut coins, i);
```
Comment (amnn) [inline crates/sui-framework/sources/pay.move:73]: Why not `pop_back` here?

Comment (damirka) [reply crates/sui-framework/sources/pay.move:73]: This PR is a copy-pasta meant to restructure stuff; but I'll look into what's happening here.

**Comment by sblackshear** [comment]
> > Suggested rule of thumb for when a function goes into `coin` vs `wallet`:
> > 
> > * If a function returns a `Coin` or `vector` of `Coin`'s, it belongs in `coin`.
> > * If it does a `transfer`, it belongs in `wallet`
> > 
> > This seems easy to apply consistently, explain in docs, and I think it will also help with naming--e.g., instead of being called `x_and_transfer`, the `coin` and `wallet function can both be called `x`.
> > @sblackshear
> 
> First I tried this approach but there are few cases such as `coin::keep` which I would personally love to keep in Coin and not require additional import for it. As for returning a `Coin` I believe that function like `divide_n` or, say, we add "split into specified amounts" should be kept in the `pay` / `wallet` module rather than in `coin` as these are too specific or "fancy".
> 
> I understand that it's not a clear distinction (although I'd prefer there was some) and suggest to follow common sense above else.

I really prefer that we would have a rule of thumb here--I think "common sense" will differ from person to person and be too hard to follow consistently. I'm not wedded to the rule I suggested, but I think we do need one.

**Comment by damirka** [comment]
@sblackshear I decided to go with your suggested rule of thumb as it's really hard to come up with something as simple as this one. Here's the result:

*sui::pay*
- now has only functions that work with `Coin`s; 
- all of the functions of the `sui::pay` are entry functions; 
- `coin::keep` is now `pay::keep` and `coin::join` is re-exported into `sui::pay` making the latter a silver bullet for wallet functionality

*sui::coin*
- contains Coin-Balance transformers (also keeps methods like `coin::put(Balance, Coin)`)
- keeps minting / burning functions as well as anything that involves TreasuryCap
- the rest of the methods are returning `Coin` (eg `divide_into_n` is moved to `coin`)

**Review by sblackshear — APPROVED** [review]
